src/Amplitude.php

```diff
@@ -474,49 +396,167 @@ public function getOptOut()
      */
-    protected function sendEvent()
+    protected function sendEvent(Event $event)
     {
-        if (empty($this->event) || empty($this->apiKey)) {
-            throw new \InternalErrorException('Event or api key not set, cannot send event');
-        }
-        $ch = curl_init(static::AMPLITUDE_API_URL);
-        if (!$ch) {
-            // Could be a number of PHP environment problems, log a critical error
-            $this->logger->critical(
-                'Call to curl_init(' . static::AMPLITUDE_API_URL . ') failed, unable to send Amplitude event'
-            );
+        if ($this->optOut) {
             return;
         }
+
+        if (!empty($this->userId)) {
+            $event->userId = $this->userId;
+        }
+
+        if (!empty($this->deviceId)) {
+            $event->deviceId = $this->deviceId;
+        }
+
+        $this->checkForApiKey();
+        $this->checkForUserIdOrDeviceId($event);
+        $ch = curl_init(static::AMPLITUDE_API_URL);
+        $this->checkCurlHandle($ch);
         $postFields = [
             'api_key' => $this->apiKey,
-            'event' => json_encode($this->event),
+            'event' => json_encode($event),
         ];
         curl_setopt($ch, \CURLOPT_POSTFIELDS, $postFields);
         // Always return instead of outputting response!
         curl_setopt($ch, \CURLOPT_RETURNTRANSFER, true);
         $response = curl_exec($ch);
-        $curlErrno = curl_errno($ch);
-        if ($curlErrno) {
-            $this->logger->critical(
-                'Curl error: ' . curl_error($ch),
-                compact('curlErrno', 'response', 'postFields')
-            );
-        } else {
-            $httpCode = curl_getinfo($ch, \CURLINFO_HTTP_CODE);
-            $this->logger->log(
-                $httpCode === 200 ? Log\LogLevel::INFO : Log\LogLevel::ERROR,
-                'Amplitude HTTP API response: ' . $response,
-                compact('httpCode', 'response', 'postFields')
-            );
-        }
+        $this->checkForCurlError($ch);
+
+        $httpCode = curl_getinfo($ch, \CURLINFO_HTTP_CODE);
+        $this->logger->log(
+            Log\LogLevel::INFO,
+            'Amplitude HTTP API response: ' . $response,
+            compact('httpCode', 'response', 'postFields')
+        );
         curl_close($ch);
     }
+
+    protected function checkForApiKey(): void
+    {
+        if (empty($this->apiKey)) {
+            throw new \LogicException(static::EXCEPTION_MSG_NO_API_KEY);
+        }
+    }
+
+    /**
+     * @param resource $ch
+     * @throws \Exception
+     */
+    protected function checkCurlHandle($ch): void
+    {
+        if (!$ch) {
+            // Could be a number of PHP environment problems, log a critical error
+            $message = 'Call to curl_init(' . static::AMPLITUDE_API_URL . ') failed, unable to send Amplitude event';
+            $this->logger->critical($message);
+            throw new \Exception($message);
```

Comment:
The current methodology of this library is to throw exceptions for anything that is a code related error (things that need to be caught hopefully before they make it to production), and log everything else.  At the end of the day this is a library for logging data for analytics on production environments.

Another way to put it, if amplitude goes down, do you want your own app to stop working completely?  Because without catching this exception, that is what would happen.

That said, using exceptions does make it easier to exit early and makes for much cleaner code in the library.  If we do change it however, we need to re-write the docs to make it clear the things that could possibly throw an exception, and make it clear that in a production env they should always wrap it in a try/catch block to prevent an outage just because amplitude goes out or something.

**tl;dr:** If we start using exceptions for things that are not code related, we need to update the documentation in the readme as well.